nossomercadolivre-api/src/main/java/com/github/nossomercadolivre/CategoryResource.java

```diff
@@ -0,0 +1,51 @@
+package com.github.nossomercadolivre;
+
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+import static com.github.nossomercadolivre.CategoryDTO.toDTO;
+import static org.springframework.http.HttpStatus.BAD_REQUEST;
+import static org.springframework.http.ResponseEntity.ok;
+
+@RestController
+@RequestMapping("/api/category")
+public class CategoryResource {
+    private final Logger log = LoggerFactory.getLogger(CategoryResource.class);
+
+    private final CategoryRepository categoryRepository;
+
+    public CategoryResource(CategoryRepository categoryRepository) {
+        this.categoryRepository = categoryRepository;
+    }
+
+    @ApiOperation(value = "Cria categoria nova")
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "", response = CategoryDTO.class),
+            @ApiResponse(code = 400, message = "", response = String.class)
+    })
+    @PostMapping()
+    public ResponseEntity<CategoryDTO> createCategory(@Valid @RequestBody final CategoryDTO categoryDTO) {
+        try {
+            Category category = categoryDTO.toModel();
```

Comment:
Aqui vou precisar que você respire e não me julgue muito hehe. O que vou sugerir não é normal, mas peço que avalie com carinho procurando os reais pontos negativos e esquecendo o que o mercado fala que é bom ou não. 

O seu categoryDTO tem tudo necessário para montar uma categoria, com ou sem mãe. A única coisa que você precisa é de um repository. Qual foi a decisão do seu código: montar a Category com os dados disponíveis dentro do DTO e o que precisava de um repositório ou qualquer outra dependência você deixou para usar do lado de fora. Em outros lugares do mercado até diriam para você usar um CategoryDTOToCategory converter. 

Qual a minha sugestão? Junte comportamento com estado. Algo assim: ```categoryDTO.toModel(categoryRepository)```. 

Você passa o que precisa como argumento e monta o objeto lá dentro. Um código que deixa claro a dependência necessária para a montagem da Categoria. Dessa forma você elimina converters, junta estado + comportamento e ainda mantem uma carga cognitiva ok do código. Tenho usado isso por muito tempo já e sinto que é o melhor meio termo até o momento. 

Claro que você pode ir além e fazer algo assim:

```
toModel(Function<Long,Category> categoryByIdLoader){
 ...
}
```
Deste jeito você minimiza o acesso dos métodos dentro do toModel e ainda pode usar o repository do lado de fora. 

O lado negativo desse approach é que você usa o controller para receber dependências que não são dele, apenas para passar para frente.

Reply:
Fica tranquilo que não vou te julgar, eu já entendi que você quer mostrar uma visão diferente de como desenvolver software melhor, estou de mente aberta, fica tranquilo :D

Inclusive, aqui https://www.infoq.com/articles/ddd-in-practice/
Tem uma implementação de referencia em java, que o método save está em uma classe com @Entity, é analogo ao que fazem em Ruby On Rails com active record.
https://guides.rubyonrails.org/active_record_basics.html

---
Bem, nesse caso não seria melhor fazer

```
CategoryDTO dto = categoryDTO.repository(categoryRespository);
dto.save();
```

Reply:
Opa, o dto se salvar seria uma opção também. Poderia ser algo assim: ```dto.save(categoryRepository)```. Não vejo lado negativo agora, o que é um problema, porque deve ter :). Só não acho interessante dividir em dois passos... Se você dividir em dois passos você abre espaço para chamar o save sem chamar o que define o repository antes :). 

*A sua api deve deixar claro o caminho que deve ser seguido pelo ponto do código que decide usá-la. Não espere que ninguém lembre de invocar nada. Faça de tudo para gerar obrigações. Quanto mais específico é seu código, menos democrático ele é.*

Reply:
Achei estranho  faazer assim
```
dto.save(categoryRepository)
```
Parece que estou salvando o categoryRepository.

Reply:
Opa, eu confesso que cada vez tenho tentado inferir responsabilidade de maneira cada vez mais lógica. Mas ainda estou no caminho. Um direcionamento(ainda não é regra) que gosto de trabalhar é: numa classe que tem estado, um método ou altera o estado dessa classe ou retorna alguma computação em função deste estado. No caso do toModel salvando, o método altera o estado de outro lugar e retorna uma computação(nova Categoria), coisa que eu não recomendo. Inclusive conseguir descrever como um dos meu pilares:
*Métodos criados nas classes que possuem os atributos que representam o estado da aplicação devem retornar algo em função do estado interno ou alterar o estado interno.*

Por conta disso, eu ainda acho que o toModel poderia simplesmente retornar a Categoria. O próximo passo do fluxo independe dele. Mas é sugestão...

Sobre o exemplo:

```
newCategoryForm.toModel(CategoryRepository :: findById)
```
A diferença é que ali na interface funcional seria ```Function<Long,Optional<Category>>```

Reply:
Acho que não estou captando a mensagem, nesse caso eu sugeriria um pair programming remoto. Se tu achar que é possível me avisa, vou mudar o código novamente para oque eu entendi agora.

Ta bem claro pra mim que esse ponto está acontecendo uma quebra no meu modelo mental  :D